src/db.rs

```diff
@@ -38,315 +40,410 @@ impl Default for Connection {
+        min_fw_api_major: policy_row.try_get::<i32, _>(2)? as u32,
+        min_fw_api_minor: policy_row.try_get::<i32, _>(3)? as u32,
+        allowed_build_ids: serde_json::from_str(&policy_row.try_get::<String, _>(4)?)?,
+    })
 }
 
-pub fn get_secret_policy(secret_id: &str) -> Result<Option<policy::Policy>> {
-    let mut dbconn = get_dbconn()?;
-    let mut trnsx = dbconn.start_transaction(TxOpts::default())?;
+pub async fn delete_policy(pid: u64) -> Result<()> {
+    let dbpool = get_dbpool().await?;
 
-    let mqstr = "SELECT polid FROM secrets WHERE polid IS NOT NULL AND secret_id = ?";
-    let policy_id: Option<u64> = trnsx.exec_first(mqstr, (secret_id,))?;
+    let query_str = "DELETE from policy WHERE id = ?";
+    let new_query_str = replace_binds(dbpool.any_kind(), query_str);
 
-    if let Some(id) = policy_id {
-        Ok(Some(get_policy(id)?))
-    } else {
-        Ok(None)
-    }
+    sqlx::query(&new_query_str)
+        .bind(pid as i64)
+        .execute(&dbpool)
+        .await?;
+    Ok(())
 }
 
-pub fn insert_keyset(ksetid: &str, kskeys: &[String], polid: Option<u32>) -> Result<()> {
-    let mut dbconn = get_dbconn()?;
-    let mut trnsx = dbconn.start_transaction(TxOpts::default())?;
+pub async fn get_secret_policy(sec: &str) -> Result<policy::Policy> {
+    let dbpool = get_dbpool().await?;
+
+    let query_str = "SELECT polid FROM secrets WHERE secret_id = ?";
+    let new_query_str = replace_binds(dbpool.any_kind(), query_str);
 
-    let mysqlstr = "INSERT INTO keysets (keysetid, kskeys, polid)
-                    VALUES(?, ?, ?)"
-        .to_string();
+    let pol_row = sqlx::query(&new_query_str)
+        .bind(sec)
+        .fetch_one(&dbpool)
+        .await?;
+    let pol = pol_row.try_get::<i64, _>(0)? as u64;
+    let secret_policy = get_policy(pol).await?;
+    Ok(secret_policy)
+}
 
-    // Create JSON for vector struct member variables
+pub async fn insert_keyset(ksetid: &str, kskeys: &[String], polid: Option<u32>) -> Result<()> {
     let kskeys_str = serde_json::to_string(kskeys)?;
 
-    trnsx.exec_drop(mysqlstr, (ksetid, &kskeys_str, polid))?;
-    trnsx.commit()?;
-    Ok(())
+    let dbpool = get_dbpool().await?;
+
+    match polid {
+        Some(p) => {
+            let query_str = "INSERT INTO keysets (keysetid, kskeys, polid) VALUES(?, ?, ?)";
+            let new_query_str = replace_binds(dbpool.any_kind(), query_str);
+            sqlx::query(&new_query_str)
+                .bind(ksetid)
+                .bind(&kskeys_str)
+                .bind(p as i64)
+                .execute(&dbpool)
+                .await?;
+            Ok(())
+        }
+        None => {
+            let query_str = "INSERT INTO keysets (keysetid, kskeys) VALUES(?, ?)";
+            let new_query_str = replace_binds(dbpool.any_kind(), query_str);
+            sqlx::query(&new_query_str)
+                .bind(ksetid)
+                .bind(&kskeys_str)
```

Comment:
again, is there any way to share the first part of this query between the two cases? Here we could also share the execution and return statement between the two cases.

Reply:
I am not sure how we can share execution and return here?  Are you thinking that we could bind the polid Option directly and if the Option value is None, then a NULL will be inserted into keysets in the polid value?  This might work, but to make it work with Any, we would have to change the option to an i64 or i32 value.  Is this something that that is important?